src/riak_kv_wm_utils.erl

```diff
@@ -335,10 +338,71 @@ is_valid_referer(RD) ->
         OriginTuple ->
             true;
         RefererTuple ->
-            lager:debug("WM referrer not origin.  Origin ~p != Referer ~p\n", [OriginTuple, RefererTuple]),
-            false
+            allowable_origin(OriginTuple, RefererTuple)
+    end.
+
+%% @doc Register allowable origins (for CORS), see riak_kv.allowable_origin
+%% w/i riak_kv.schema.
+-spec register_allowable_origins(list()) -> ok.
+register_allowable_origins(Origins) ->
+    maybe_create_allowable_origins_ets(),
+    register_allowable_origins1(Origins).
+
+register_allowable_origins1([]) -> ok;
+register_allowable_origins1([Origin = '*'|_T]) ->
+    Origin1 = normalize_referer(Origin),
+    %% wrap the origin for full match on ets lookup
+    ets:insert(?ALLOWABLE_ORIGINS_TABLE, {Origin1});
+register_allowable_origins1([http, HttpOrigins|T]) ->
+    register_allowable_origins2(http, HttpOrigins),
+    register_allowable_origins1(T);
+register_allowable_origins1([https, HttpOrigins|T]) ->
+    register_allowable_origins2(https, HttpOrigins),
+    register_allowable_origins1(T).
+
+register_allowable_origins2(Scheme, HostPortPairs) ->
+    OriginsWild = [{normalize_referer({Scheme, Host, "*"})} || {Host} <- HostPortPairs],
+    Origins = [{normalize_referer({Scheme, Host, Port})} || {Host, Port} <- HostPortPairs],
+    ets:insert(?ALLOWABLE_ORIGINS_TABLE, OriginsWild ++ Origins).
+
+allowable_origins_table_exists() ->
+    undefined /= ets:info(?ALLOWABLE_ORIGINS_TABLE).
+
+maybe_create_allowable_origins_ets() ->
+    case allowable_origins_table_exists() of
+        false ->
+            ets:new(?ALLOWABLE_ORIGINS_TABLE,
+                    [named_table, public, ordered_set]);
+        _ ->
+            ok
     end.
 
+allowable_origin(OriginTuple, RefererTuple) ->
+    RefererTuple1 = {Scheme, Host, Port} = normalize_referer(RefererTuple),
+    case allowable_origins_table_exists() andalso
+        (ets:member(?ALLOWABLE_ORIGINS_TABLE, RefererTuple1) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, Host, "*"}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, "*", "*"}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {Scheme, "*", Port}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {"*", Host, Port}) orelse
+         ets:member(?ALLOWABLE_ORIGINS_TABLE, {"*", "*", "*"})) of
+        false ->
+            lager:debug("WM referrer not origin. Origin ~p != Referer ~p" ++
+                        " and Referer not in allowable origins.\n",
+                        [OriginTuple, RefererTuple]),
+            false;
+        _ -> true
+    end.
+
+normalize_referer('*') -> {"*", "*", "*"};
+normalize_referer("*") -> normalize_referer('*');
```

Comment:
As noted in the main PR comments, this function head is irrelevant and giving dialyzer heartburn. Same with the one that follows.